pypulseq/Sequence/sequence.py

```diff
@@ -611,67 +611,105 @@ def gradient_waveforms(self) -> np.ndarray:
-                        waveform = grad.waveform
-                    else:
-                        nt_start = np.round(grad.delay / self.grad_raster_time)
-                        if np.abs(grad.flat_time) > eps:
-                            t = np.cumsum(
-                                [0, grad.rise_time, grad.flat_time, grad.fall_time]
-                            )
-                            trap_form = np.array([0, 1, 1, 0]) * grad.amplitude
-                        else:
-                            t = np.cumsum([0, grad.rise_time, grad.fall_time])
-                            trap_form = np.array([0, 1, 0]) * grad.amplitude
-
-                        tn = math.floor(t[-1] / self.grad_raster_time)
-                        t = np.append(t, t[-1] + self.grad_raster_time)
-                        trap_form = np.append(trap_form, 0)
-
-                        if np.abs(grad.amplitude) > eps:
-                            waveform = points_to_waveform(
-                                times=t,
-                                amplitudes=trap_form,
-                                grad_raster_time=self.grad_raster_time,
-                            )
-                        else:
-                            waveform = np.zeros(tn + 1)
-
-                    if len(waveform) != np.sum(np.isfinite(waveform)):
-                        warn("Not all elements of the generated waveform are finite")
-
-                    """
-                    Matlab dynamically resizes arrays during slice assignment operation if assignment is out of bounds
-                    Numpy does not; following is a workaround
-                    """
-                    l1, l2 = int(t0_n + nt_start), int(t0_n + nt_start + len(waveform))
-                    if l2 > grad_waveforms.shape[1]:
-                        z = np.zeros(
-                            (grad_waveforms.shape[0], l2 - grad_waveforms.shape[1])
-                        )
-                        grad_waveforms = np.hstack((grad_waveforms, z))
-                    grad_waveforms[j, l1:l2] = waveform
+        dt = self.system.grad_raster_time
 
-            t0 += self.block_durations[block_counter]
-            t0_n = np.round(t0 / self.grad_raster_time)
+        tx = ws[0][0]
+        ty = ws[1][0]
+        tz = ws[2][0]
+
+        gx = points_to_waveform(times=tx, amplitudes=ws[0][1], grad_raster_time=dt)
+        gy = points_to_waveform(times=ty, amplitudes=ws[1][1], grad_raster_time=dt)
+        gz = points_to_waveform(times=tz, amplitudes=ws[2][1], grad_raster_time=dt)
+
+        # Check for empty arrays, end set end and starting times
+        tx_s = 0 if tx.size==0 else tx[0]
+        ty_s = 0 if ty.size==0 else ty[0]
+        tz_s = 0 if tz.size==0 else tz[0]
+
+        tx_e = 0 if tx.size==0 else tx[-1]
+        ty_e = 0 if ty.size==0 else ty[-1]
+        tz_e = 0 if tz.size==0 else tz[-1]
 
+        max_t = round(max(tx_e, ty_e, tz_e)/dt)*dt

```

Comment:
Request all `round`, `min` and `max` built-in functions to be refactored to `numpy` functions. I've tried to do this throughout the repo for consistency.